python/ray/data/_internal/util.py

```diff
@@ -51,3 +58,94 @@ def _check_pyarrow_version():
             )
         else:
             _VERSION_VALIDATED = True
+
+
+def _autodetect_parallelism(
+    parallelism: int,
+    cur_pg: Optional["PlacementGroup"],
+    reader: Optional["Reader"] = None,
+    avail_cpus: Optional[int] = None,
+) -> (int, int):
+    """Returns parallelism to use and the min safe parallelism to avoid OOMs.
+
+    This detects parallelism based on three heuristic functions:
+
+     1) Available CPUs. We detect how many CPUs are available for use in the cluster,
+        and ensure the parallelism is high enough to make use of the CPUs.
+     2) Min block size. We try to avoid creating blocks smaller than this threshold,
+        to avoid the overhead of tiny blocks.
+     3) Max block size. We try to avoid creating blocks larger than this threshold,
+        since it can lead to OOM errors during processing.
+
+    Args:
+        parallelism: The user-requested parallelism, or -1 for auto-detection.
+        cur_pg: The current placement group, to be used for avail cpu calculation.
+        reader: The datasource reader, to be used for data size estimation.
+        avail_cpus: Override avail cpus detection (for testing only).
+
+    Returns:
+        Tuple of detected parallelism (only if -1 was specified), and the min safe
+        parallelism (which can be used to generate warnings about large blocks).
+    """
+    min_safe_parallelism = 1
+    ctx = DatasetContext.get_current()
+    if reader:
+        mem_size = reader.estimate_inmemory_data_size()
+        if mem_size is not None:
+            min_safe_parallelism = max(1, int(mem_size / ctx.target_max_block_size))
+            max_reasonable_parallelism = max(
+                1, int(mem_size / ctx.target_min_block_size)
+            )
+    else:
+        mem_size = None
+        max_reasonable_parallelism = 9999999
+    if parallelism < 0:
+        if parallelism != -1:
+            raise ValueError("`parallelism` must either be -1 or a positive integer.")
+        # Start with 2x the number of cores as a baseline, with a min floor.
+        avail_cpus = avail_cpus or _estimate_avail_cpus(cur_pg)
+        parallelism = max(
+            min(ctx.min_parallelism, max_reasonable_parallelism),
```

Comment:
tbh i know this function is developed as-is today probably due to many legacy reasons but i found this function very critical for perf but also quite hard to read between all the min/max and nested min/max .. It looks high ROI to me to refactor this function completely into more modular functions, each covered by its own test so this decision making is clear for everyone else as a straight set of function calls for each rule, what do you think ?

Reply:
Nack. We have a test added that covers this function in all the edge scenarios... I think that is more than enough. It's not useful to test them in smaller units since it's the interactions that are important.